File: dotnet/Allmark/RenderConsole/ConsoleFootnoteListRenderer.cs
namespace Allmark.Render;

using Allmark.Types;

public static class ConsoleFootnoteListRenderer
{
    public static OutputRenderer Create()
    {
        return new OutputRenderer
        {
            Name = "footnote_list",
            Render = (node, state, decode) => Render(node, state),
        };
    }

    public static void Render(MarkdownNode node, RendererState state)
    {
        if (state.Footnotes.Count == 0)
        {
            return;
        }
        state.Output.Append($"\n{Ansi.Dim}---{Ansi.Reset}\n");
        var number = 1;
        foreach (var footnote in state.Footnotes)
        {
            var label = number++;
            state.Output.Append($"{Ansi.Dim}[{label}]{Ansi.Reset} ");
            if (footnote.Info != null && state.FootnoteRefs.TryGetValue(footnote.Info, out var refNode))
            {
                RenderChildren.Execute(refNode, state);
            }
            var output = state.Output.ToString();
            if (output.EndsWith("\n"))
            {
                state.Output.Length -= 1;
            }
            state.Output.Append("\n");
        }
    }
}

File: zig/src/render-console/consoleFootnoteListRenderer.zig
const std = @import("std");
const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const RendererState = @import("../types/RendererState.zig").RendererState;
const Renderer = @import("../types/Renderer.zig").Renderer;
const renderChildrenFn = @import("console.zig").renderChildrenConsole;
const ansiDim = @import("console.zig").ansiDim;
const ansiReset = @import("console.zig").ansiReset;

pub fn render(node: *const MarkdownNode, state: *RendererState, decode: ?bool) void {
    _ = node;
    _ = decode;
    if (state.footnotes.items.len == 0) {
        return;
    }

    state.output.appendSlice(state.allocator, "\n") catch unreachable;
    state.output.appendSlice(state.allocator, ansiDim) catch unreachable;
    state.output.appendSlice(state.allocator, "---") catch unreachable;
    state.output.appendSlice(state.allocator, ansiReset) catch unreachable;
    state.output.appendSlice(state.allocator, "\n") catch unreachable;

    var number: usize = 1;
    for (state.footnotes.items) |fn_node| {
        const label = number;
        number += 1;

        const label_str = std.fmt.allocPrint(state.allocator, "[{d}]", .{label}) catch unreachable;
        defer state.allocator.free(label_str);

        state.output.appendSlice(state.allocator, ansiDim) catch unreachable;
        state.output.appendSlice(state.allocator, label_str) catch unreachable;
        state.output.appendSlice(state.allocator, ansiReset) catch unreachable;
        state.output.appendSlice(state.allocator, " ") catch unreachable;

        try renderChildrenFn(fn_node, state, true);

        if (state.output.items.len > 0 and state.output.items[state.output.items.len - 1] == '\n') {
            state.output.shrinkRetainingCapacity(state.output.items.len - 1);
        }
        if (state.output.items.len > 0 and state.output.items[state.output.items.len - 1] == '\r') {
            state.output.shrinkRetainingCapacity(state.output.items.len - 1);
        }

        state.output.appendSlice(state.allocator, "\n") catch unreachable;
    }
}

pub const consoleFootnoteListRenderer = Renderer{
    .name = "footnote_list",
    .render = render,
};
